### PR TITLE
check for API -- WIP

### PR DESCRIPTION
**Pull request by jbenet**
This is a WIP to make the ipfs cli tool check for a running daemon
via the API, and make sure the APIs match.

I ran into some complexity i need to think more about before
attepting to fix... anyone feel free to pick this up!

License: MIT
Signed-off-by: Juan Batiz-Benet <juan@benet.ai>

### CODE REVIEW

File: cmd/ipfs/main.go [hunk 1]
```diff
@@ -32,8 +34,12 @@ import (
 // log is the command logger
 var log = eventlog.Logger("cmd/ipfs")
 
-// signal to output help
-var errHelpRequested = errors.New("Help Requested")
+var (
+	errHelpRequested       = errors.New("Help Requested")
```
Comment (whyrusleeping) [inline cmd/ipfs/main.go]: this error isnt even used...

Comment (jbenet) [inline cmd/ipfs/main.go]: ok lets kill it then :)

Comment (whyrusleeping) [inline cmd/ipfs/main.go]: i love removing code...

**Comment by jbenet** [comment]
This will be hard to test right-- a few cases:
- check things work ok offline + without $repo/api file
- check starting daemon places the $repo/api file
- check client reads it and uses it ok
- check client obeys it (change it to some other value + nc)
- check client correctly errors if api is off (api file different), but daemon is on (repo locked)
- check client works ok if daemon is turned off, even with $repo/api file there

**Comment by jbenet** [comment]
@rht or @chriscool could i get help testing o/ ? i've very little time this week, but this is an important thing to merge soon

**Comment by rht** [comment]
>￼ check client correctly errors if api is off (api file different), but daemon is on (repo locked)

There are 3 addressed involved: client addr, api file addr, server addr.
1. c = f, f != s
2. c = s, f != s
3. f = s, c != f
4. c = f = s
5. c != f, c != s, f != s

**Comment by jbenet** [comment]
by "api is off" i mean that the repo is locked, and the `/api` file is there, but the address on the `/api` file is wrong-- or nothing is there. (i.e. testing that the cli errors correctly if the api endpoint cannot be reached)

**Comment by rht** [comment]
> but the address on the /api file is wrong-- or nothing is there

Are these cases covered by a test that simply check if the api file is different from the api specified in `ipfs-client --api`?
https://github.com/ipfs/go-ipfs/commit/dd265b7f034488b6b4056a061276efe6baaeb530#diff-21a8cbf17e33b82a8fca825aca160f1dR25

I put 4 cases. I think they cover all the checklists (the unchecked ones indirectly).

**Comment by jbenet** [comment]
@rht sorry i commented on the commit instead of the PR: https://github.com/ipfs/go-ipfs/commit/dd265b7f034488b6b4056a061276efe6baaeb530

**Comment by chriscool** [comment]
About test_lauch_ipfs_daemon() arguments, my prefered solution would be something like this:

https://github.com/chriscool/go-ipfs/commit/e1d8200ffb83a863142419ce796de948d7cd34c0

**Comment by jbenet** [comment]
@chriscool's solution LGTM

File: cmd/ipfs/main.go [hunk 2]
```diff
@@ -32,8 +34,10 @@ import (
 // log is the command logger
 var log = eventlog.Logger("cmd/ipfs")
 
-// signal to output help
-var errHelpRequested = errors.New("Help Requested")
+var (
+	errUnexpectedApiOutput = errors.New("api returned unexpected output")
+	errApiVersionMismatch  = errors.New("api version mismatch")
+)
```
Comment (rht) [inline cmd/ipfs/main.go]: can't comment below, but from varcheck (https://github.com/opennota/check), L46 is unused.

**Comment by jbenet** [comment]
@rht 
> I find saying 'no api' when it actually is 'different api' confusing. It just happens in this case they both have the same err.

What i mean by "no api" is that the endpoint does not exist. an HTTP request will fail. that's ultimately what drives the error, not the other stuff.

**Comment by jbenet** [comment]
thanks for helping me with the tests @rht !

**Comment by jbenet** [comment]
This PR is a big deal. you can now very easily do:

```
ipfs --api <addr-of-remote-endpoint>
```
and it should "just work". this was one of the most requested features.

cc a few people i remember asked for it: @whyrusleeping  @diasdavid @krl @lgierth @wking @ehd @cryptix @Luzifer

**Comment by diasdavid** [comment]
woot! :D

**Pull request merged**